File: src/com/twicestruck/jqc/vm/Instructions.java
package com.twicestruck.jqc.vm;

public class Instructions
{
	public static final byte DONE       = 0x0000;

	public static final byte MUL_F      = 0x0001;
	public static final byte MUL_V      = 0x0002;
	public static final byte MUL_FV     = 0x0003;
	public static final byte MUL_VF     = 0x0004;
	public static final byte DIV_F      = 0x0005;
	public static final byte ADD_F      = 0x0006;
	public static final byte ADD_V      = 0x0007;
	public static final byte SUB_F      = 0x0008;
	public static final byte SUB_V      = 0x0009;

	public static final byte EQ_F       = 0x000A;
	public static final byte EQ_V       = 0x000B;
	public static final byte EQ_S       = 0x000C;
	public static final byte EQ_E       = 0x000D;
	public static final byte EQ_FNC     = 0x000E;

	public static final byte NE_F       = 0x000F;
	public static final byte NE_V       = 0x0010;
	public static final byte NE_S       = 0x0011;
	public static final byte NE_E       = 0x0012;
	public static final byte NE_FNC     = 0x0013;

	public static final byte LE         = 0x0014;
	public static final byte GE         = 0x0015;
	public static final byte LT         = 0x0016;
	public static final byte GT         = 0x0017;

	public static final byte LOAD_F     = 0x0018;
	public static final byte LOAD_V     = 0x0019;
	public static final byte LOAD_S     = 0x001A;
	public static final byte LOAD_ENT   = 0x001B;
	public static final byte LOAD_FLD   = 0x001C;
	public static final byte LOAD_FNC   = 0x001D;

	public static final byte ADDRESS    = 0x001E;

	public static final byte STORE_F    = 0x001F;
	public static final byte STORE_V    = 0x0020;
	public static final byte STORE_S    = 0x0021;
	public static final byte STORE_ENT  = 0x0022;
	public static final byte STORE_FLD  = 0x0023;
	public static final byte STORE_FNC  = 0x0024;

	public static final byte STOREP_F   = 0x0025;
	public static final byte STOREP_V   = 0x0026;
	public static final byte STOREP_S   = 0x0027;
	public static final byte STOREP_ENT = 0x0028;
	public static final byte STOREP_FLD = 0x0029;
	public static final byte STOREP_FNC = 0x002A;

	public static final byte RETURN     = 0x002B;

	public static final byte NOT_F      = 0x002C;
	public static final byte NOT_V      = 0x002D;
	public static final byte NOT_S      = 0x002E;
	public static final byte NOT_ENT    = 0x002F;
	public static final byte NOT_FNC    = 0x0030;

	public static final byte IF         = 0x0031;
	public static final byte IFNOT      = 0x0032;

	public static final byte CALL0      = 0x0033;
	public static final byte CALL1      = 0x0034;
	public static final byte CALL2      = 0x0035;
	public static final byte CALL3      = 0x0036;
	public static final byte CALL4      = 0x0037;
	public static final byte CALL5      = 0x0038;
	public static final byte CALL6      = 0x0039;
	public static final byte CALL7      = 0x003A;
	public static final byte CALL8      = 0x003B;

	public static final byte STATE      = 0x003C;

	public static final byte GOTO       = 0x003D;

	public static final byte AND        = 0x003E;
	public static final byte OR         = 0x003F;

	public static final byte BITAND     = 0x0040;
	public static final byte BITOR      = 0x0041;

	// these aren't instructions, but shorthand for the range
	public static final byte MIN        = 0x0000;
	public static final byte MAX        = 0x0041;
}

File: kzqcvm/instructions.cpp
/*
Kzqcvm QuakeC VM Interpreter
Copyright (c) 2010 David Laurie

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
/*
kzqcvm/instructions.cpp
*/

#include "instructions.h"

#include <stdint.h>

//-----------------------------------------------------------------------------
namespace kzqcvm {
//-----------------------------------------------------------------------------

InstructionInfo instruction_info[] = {
	{ Instructions::DONE,       "DONE",       { IT_VECTOR,   IT_NONE,     IT_NONE     } },

	{ Instructions::MUL_F,      "MUL_F",      { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::MUL_V,      "MUL_V",      { IT_VECTOR,   IT_VECTOR,   IT_FLOAT    } },
	{ Instructions::MUL_FV,     "MUL_FV",     { IT_FLOAT,    IT_VECTOR,   IT_FLOAT    } },
	{ Instructions::MUL_VF,     "MUL_VF",     { IT_VECTOR,   IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::DIV_F,      "DIF_F",      { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::ADD_F,      "ADD_F",      { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::ADD_V,      "ADD_V",      { IT_VECTOR,   IT_VECTOR,   IT_VECTOR   } },
	{ Instructions::SUB_F,      "SUB_F",      { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::SUB_V,      "SUB_V",      { IT_VECTOR,   IT_VECTOR,   IT_VECTOR   } },

	{ Instructions::EQ_F,       "EQ_F",       { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::EQ_V,       "EQ_V",       { IT_VECTOR,   IT_VECTOR,   IT_FLOAT    } },
	{ Instructions::EQ_S,       "EQ_S",       { IT_STRING,   IT_STRING,   IT_FLOAT    } },
	{ Instructions::EQ_E,       "EQ_E",       { IT_ENTITY,   IT_ENTITY,   IT_FLOAT    } },
	{ Instructions::EQ_FNC,     "EQ_FNC",     { IT_FUNCTION, IT_FUNCTION, IT_FLOAT    } },

	{ Instructions::NE_F,       "NE_F",       { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::NE_V,       "NE_V",       { IT_VECTOR,   IT_VECTOR,   IT_FLOAT    } },
	{ Instructions::NE_S,       "NE_S",       { IT_STRING,   IT_STRING,   IT_FLOAT    } },
	{ Instructions::NE_E,       "NE_E",       { IT_ENTITY,   IT_ENTITY,   IT_FLOAT    } },
	{ Instructions::NE_FNC,     "NE_FNC",     { IT_FUNCTION, IT_FUNCTION, IT_FLOAT    } },

	{ Instructions::LE,         "LE",         { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::GE,         "GE",         { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::LT,         "LT",         { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::GT,         "GT",         { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },

	{ Instructions::LOAD_F,     "LOAD_F",     { IT_ENTITY,   IT_FIELD,    IT_FLOAT    } },
	{ Instructions::LOAD_V,     "LOAD_V",     { IT_ENTITY,   IT_FIELD,    IT_VECTOR   } },
	{ Instructions::LOAD_S,     "LOAD_S",     { IT_ENTITY,   IT_FIELD,    IT_STRING   } },
	{ Instructions::LOAD_ENT,   "LOAD_ENT",   { IT_ENTITY,   IT_FIELD,    IT_ENTITY   } },
	{ Instructions::LOAD_F,     "LOAD_FLD",   { IT_ENTITY,   IT_FIELD,    IT_FIELD    } },
	{ Instructions::LOAD_F,     "LOAD_FNC",   { IT_ENTITY,   IT_FIELD,    IT_FUNCTION } },

	{ Instructions::ADDRESS,    "ADDRESS",    { IT_ENTITY,   IT_FIELD,    IT_ADDRESS  } },

	{ Instructions::STORE_F,    "STORE_F",    { IT_FLOAT,    IT_FLOAT,    IT_NONE     } },
	{ Instructions::STORE_V,    "STORE_V",    { IT_VECTOR,   IT_VECTOR,   IT_NONE     } },
	{ Instructions::STORE_S,    "STORE_S",    { IT_STRING,   IT_STRING,   IT_NONE     } },
	{ Instructions::STORE_ENT,  "STORE_ENT",  { IT_ENTITY,   IT_ENTITY,   IT_NONE     } },
	{ Instructions::STORE_FLD,  "STORE_FLD",  { IT_FIELD,    IT_FIELD,    IT_NONE     } },
	{ Instructions::STORE_FNC,  "STORE_FNC",  { IT_FUNCTION, IT_FUNCTION, IT_NONE     } },

	{ Instructions::STOREP_F,   "STOREP_F",   { IT_FLOAT,    IT_ADDRESS,  IT_NONE     } },
	{ Instructions::STOREP_V,   "STOREP_V",   { IT_VECTOR,   IT_ADDRESS,  IT_NONE     } },
	{ Instructions::STOREP_S,   "STOREP_S",   { IT_STRING,   IT_ADDRESS,  IT_NONE     } },
	{ Instructions::STOREP_ENT, "STOREP_ENT", { IT_ENTITY,   IT_ADDRESS,  IT_NONE     } },
	{ Instructions::STOREP_FLD, "STOREP_FLD", { IT_FIELD,    IT_ADDRESS,  IT_NONE     } },
	{ Instructions::STOREP_FNC, "STOREP_FNC", { IT_FUNCTION, IT_ADDRESS,  IT_NONE     } },

	{ Instructions::RETURN,     "RETURN",     { IT_VECTOR,   IT_NONE,     IT_NONE     } },

	{ Instructions::NOT_F,      "NOT_F",      { IT_FLOAT,    IT_NONE,     IT_FLOAT    } },
	{ Instructions::NOT_V,      "NOT_V",      { IT_VECTOR,   IT_NONE,     IT_FLOAT    } },
	{ Instructions::NOT_S,      "NOT_S",      { IT_STRING,   IT_NONE,     IT_FLOAT    } },
	{ Instructions::NOT_ENT,    "NOT_ENT",    { IT_ENTITY,   IT_NONE,     IT_FLOAT    } },
	{ Instructions::NOT_FNC,    "NOT_FNC",    { IT_FUNCTION, IT_NONE,     IT_FLOAT    } },

	{ Instructions::IF,         "IF",         { IT_FLOAT,    IT_DIRECT,   IT_NONE     } },
	{ Instructions::IFNOT,      "IFNOT",      { IT_FLOAT,    IT_DIRECT,   IT_NONE     } },

	{ Instructions::CALL0,      "CALL0",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },
	{ Instructions::CALL1,      "CALL1",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },
	{ Instructions::CALL2,      "CALL2",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },
	{ Instructions::CALL3,      "CALL3",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },
	{ Instructions::CALL4,      "CALL4",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },
	{ Instructions::CALL5,      "CALL5",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },
	{ Instructions::CALL6,      "CALL6",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },
	{ Instructions::CALL7,      "CALL7",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },
	{ Instructions::CALL8,      "CALL8",      { IT_FUNCTION, IT_NONE,     IT_NONE     } },

	{ Instructions::STATE,      "STATE",      { IT_FLOAT,    IT_FUNCTION, IT_NONE     } },

	{ Instructions::GOTO,       "GOTO",       { IT_DIRECT,   IT_NONE,     IT_NONE     } },

	{ Instructions::AND,        "AND",        { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::OR,         "OR",         { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },

	{ Instructions::BITAND,     "BITAND",     { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },
	{ Instructions::BITOR,      "BITOR",      { IT_FLOAT,    IT_FLOAT,    IT_FLOAT    } },

	{ 0x000,                    "UNKNOWN",    { IT_NONE,     IT_NONE,     IT_NONE     } }
};

const int num_instruction_info = sizeof(instruction_info) / sizeof(InstructionInfo);

InstructionInfo *GetInstructionInfo(int16_t instruction)
{
	for (int i=0; i<num_instruction_info; ++i)
	{
		if (instruction_info[i].value == instruction)
		{
			return &instruction_info[i];
		}
	}
	return &instruction_info[num_instruction_info-1];
}

const char *GetInstructionName(int16_t instruction)
{
	return GetInstructionInfo(instruction)->name;
}

//-----------------------------------------------------------------------------
} // namespace
//-----------------------------------------------------------------------------
